### PR TITLE
feat(pruner, primitives): move prune batch sizes to `ChainSpec`

### PR DESCRIPTION
**Pull request by shekhirin**
We need to set pruner batch sizes depending on the network the node is running. The reason for this is that mainnet has more transactions per block than testnets, hence more receipts, tx senders, tx lookup entries and changesets are inserted in the database. We need to prune more data than the amount of new incoming data, but also not too much because otherwise the problems described in https://github.com/paradigmxyz/reth/pull/4246 will arise.

### CODE REVIEW

File: crates/primitives/src/prune/batch_sizes.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+impl PruneBatchSizes {
+    /// Default prune batch sizes for Ethereum mainnet.
+    /// These settings are sufficient to prune more data than generated with each new block.
+    pub const fn mainnet() -> Self {
+        Self {
+            receipts: 250,
+            transaction_lookup: 250,
+            transaction_senders: 250,
+            account_history: 1000,
```
Comment (gakonst) [inline crates/primitives/src/prune/batch_sizes.rs:61]: assuming these are heuristics/made up?

File: crates/primitives/src/prune/batch_sizes.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    storage_history: usize,
+}
+
+macro_rules! impl_prune_batch_size_methods {
+    ($(($human_name:expr, $name:ident)),+) => {
+        paste! {
+            impl PruneBatchSizes {
+                $(
+                    #[doc = concat!("Maximum number of ", $human_name, " to prune, accounting for the block interval.")]
+                    pub fn $name(&self, block_interval: usize) -> usize {
```
Comment (gakonst) [inline crates/primitives/src/prune/batch_sizes.rs:30]: `receipts_to_prune`? or something? `receipts` on its own seems weird

Comment (gakonst) [inline crates/primitives/src/prune/batch_sizes.rs:30]: `receipts_to_prune`? or something? `receipts` on its own seems weird

File: crates/primitives/src/chain/spec.rs [hunk 3]
```diff
@@ -63,7 +63,8 @@ pub static MAINNET: Lazy<Arc<ChainSpec>> = Lazy::new(|| {
             11052984,
             H256(hex!("649bbc62d0e31342afea4e5cd82d4049e7e1ee912fc0889aa790803be39038c5")),
         )),
-        ..Default::default()
+        base_fee_params: BaseFeeParams::ethereum(),
+        prune_batch_sizes: PruneBatchSizes::mainnet(),
```
Comment (gakonst) [inline crates/primitives/src/chain/spec.rs:67]: nice

**Review by gakonst — APPROVED** [review]
fine w me

File: crates/primitives/src/prune/batch_sizes.rs (hunk 1 above)
Comment (shekhirin) [reply crates/primitives/src/prune/batch_sizes.rs:61]: - `receipts`, `transaction_lookup` and `transaction_senders` – basically transactions per block which currently sits at 150 avg, occasionally jumping up to 250
- `account_history` – `AccountChangeSet` table on avg grows for 200 entries per block, `AccountHistory` is cleared up based on the results of clearing the changesets
- `storage_history` – `StorageChangeSet` table on avg grows for 500 entries per block, `StorageHistory` is cleared up based on the results of clearing the changesets


I just added a new DB metric https://github.com/paradigmxyz/reth/pull/4316 which will allow to measure these values more accurately, as I was gathering the data just snapshotting the `db stats`

**Review by joshieDo — APPROVED** [review]
lgtm,

maybe it the future it can be self adaptive, since i guess it also depends on the machine  running it

how did it perform on the last run? 
- did it start losing track of the tip? reduce batch size.
- did it have "time to spare"? increase batch size

**Comment by shekhirin** [comment]
> since i guess it also depends on the machine running it

I'm not sure there's a machine that can keep up with the tip and at the same time run the pruner for too long with current settings. 

> did it start losing track of the tip? reduce batch size.
did it have "time to spare"? increase batch size

yep, something like this should work. We can also be dependent on the number of transactions in the previous block for receipts/tx senders/tx lookup.

**Pull request merged**